### ISSUE TITLE
Invalid pointer free when error occurs inside of callback.msiSetKeyValuePairsToObj

### ISSUE THREAD

**Issue by alanking** [irods/irods_rule_engine_plugin_python#96]
- [x] main
- [x] 4-2-stable
---------

### iRODS Version, OS and Version
4.2.11 server, 4.2.11.0 plugin

### What did you try to do?
Execute a dynamic PEP which calls a microservice from a configured rulebase in PREP (e.g. core.py). The rule looks like this:
```python
def pep_api_touch_pre(rule_args, callback, rei):
    out = callback.msiString2KeyValPair('touch_pre_fired=YES', 0)
    if not out['status']:
        callback.writeLine('stdout', 'ERROR: ' + str(out['code']))
    kvp = out['arguments'][1]
    callback.msiSetKeyValuePairsToObj(kvp, '{0}', '-d') # this should produce an error
```

This can be triggered by simply executing `itouch`:
```bash
$ itouch foo
```

### Expected behavior
An error code and python stacktrace in the log, like this:
```
Dec 21 17:00:21 pid:3215 NOTICE: checkAndGetObjectId cmlExecuteNoAnswerSql(rollback) succeeded
Dec 21 17:00:21 pid:3215 NOTICE: rsModAVUMetadata: rcModAVUMetadata failed
Dec 21 17:00:21 pid:3215 remote addresses: 192.168.16.2, 192.168.16.3 ERROR: caught python exception: Traceback (most recent call last):
  File "/etc/irods/core.py", line 337, in pep_api_touch_pre
    callback.msiSetKeyValuePairsToObj(kvp, '{0}', '-d')
RuntimeError: [iRods__Error__Code:-817000] [-]  /irods_source/server/re/include/irods_re_plugin.hpp:326:irods::error irods::dynamic_operation_execution_manager<std::__1::tuple<>, RuleExecInfo *, irods::DONT_AUDIT_RULE>::call(std::string, std::string, OP, As &&...) [T = std::__1::tuple<>, C = RuleExecInfo *, Audit = irods::DONT_AUDIT_RULE, OP = std::__1::function<irods::error (const std::__1::basic_string<char> &, RuleExecInfo *&, irods::unpack &&)>, As = <const std::__1::basic_string<char> &, RuleExecInfo *&, irods::unpack>] :  status [CAT_UNKNOWN_FILE]  errno [] -- message [exec_microservice_adapter failed]
        [-]     /irods_source/server/re/src/irods_re_plugin.cpp:132:irods::error irods::default_microservice_manager<RuleExecInfo *>::exec_microservice_adapter(std::string, irods::default_ms_ctx, std::list<boost::any> &) :  status [CAT_UNKNOWN_FILE]  errno [] -- message [exec_microservice_adapter failed]
```

### Observed behavior (including steps to reproduce, if applicable)
An invalid pointer free kills the connected agent. A sample of the output in the log:
```
Dec 21 17:29:21 pid:3419 NOTICE: rsModAVUMetadata: rcModAVUMetadata failed
free(): invalid pointer
Caught signal [6]. Dumping stacktrace and exiting

Dumping stack trace
<0>     Offset:         Address: 0x7fe183b8c9b8 /usr/lib/libirods_server.so.4.2.11(+0x15149b8) [0x7fe183b8c9b8]
<1>     Offset:         Address: 0x7fe18095a980 /lib/x86_64-linux-gnu/libpthread.so.0(+0x12980) [0x7fe18095a980]
<2>     Offset:         Address: 0x7fe180595fb7 /lib/x86_64-linux-gnu/libc.so.6(gsignal+0xc7) [0x7fe180595fb7]
<3>     Offset:         Address: 0x7fe180597921 /lib/x86_64-linux-gnu/libc.so.6(abort+0x141) [0x7fe180597921]
<4>     Offset:         Address: 0x7fe1805e0967 /lib/x86_64-linux-gnu/libc.so.6(+0x89967) [0x7fe1805e0967]
<5>     Offset:         Address: 0x7fe1805e79da /lib/x86_64-linux-gnu/libc.so.6(+0x909da) [0x7fe1805e79da]
<6>     Offset:         Address: 0x7fe1805eef0c /lib/x86_64-linux-gnu/libc.so.6(cfree+0x4dc) [0x7fe1805eef0c]
<7>     Offset:         Address: 0x7fe184f82222 /usr/lib/libirods_common.so.4.2.11(clearMsParam+0x92) [0x7fe184f82222]
<8>     Offset:         Address: 0x7fe17ce163b4 /usr/lib/irods/plugins/rule_engines/libirods_rule_engine_plugin-python.so(+0x75f3b4) [0x7fe17ce163b4]
<9>     Offset: 0x76    Address: 0x7fe17ce2c206 boost::python::detail::raw_dispatcher<boost::python::dict (*)(boost::python::tuple const&, boost::python::dict const&)>::operator()(_object*, _object*)
<10>    Offset: 0x34    Address: 0x7fe17ce2c104 boost::python::objects::full_py_function_impl<boost::python::detail::raw_dispatcher<boost::python::dict (*)(boost::python::tuple const&, boost::python::dict const&)>, boost::mpl::vector1<_object*> >::operator()(_object*, _object*)
<11>    Offset: 0x2fb   Address: 0x7fe17c48f85b boost::python::objects::function::call(_object*, _object*) const
<12>    Offset:         Address: 0x7fe17c491ad5 /opt/irods-externals/boost1.67.0-0/lib/libboost_python27.so.1.67.0(+0x25ad5) [0x7fe17c491ad5]
<13>    Offset: 0x47    Address: 0x7fe17c498e87 boost::python::handle_exception_impl(boost::function0<void>)
<14>    Offset:         Address: 0x7fe17c491766 /opt/irods-externals/boost1.67.0-0/lib/libboost_python27.so.1.67.0(+0x25766) [0x7fe17c491766]
<15>    Offset:         Address: 0x7fe17c063913 /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyObject_Call+0x43) [0x7fe17c063913]
<16>    Offset:         Address: 0x7fe17c090eec /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(+0x1a1eec) [0x7fe17c090eec]
<17>    Offset:         Address: 0x7fe17c063913 /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyObject_Call+0x43) [0x7fe17c063913]
<18>    Offset:         Address: 0x7fe17c00f31a /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(+0x12031a) [0x7fe17c00f31a]
<19>    Offset:         Address: 0x7fe17c063913 /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyObject_Call+0x43) [0x7fe17c063913]
<20>    Offset:         Address: 0x7fe17bf8d951 /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyEval_EvalFrameEx+0x41e1) [0x7fe17bf8d951]
<21>    Offset:         Address: 0x7fe17c0ce908 /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyEval_EvalCodeEx+0x7d8) [0x7fe17c0ce908]
<22>    Offset:         Address: 0x7fe17c069649 /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(+0x17a649) [0x7fe17c069649]
<23>    Offset:         Address: 0x7fe17c063913 /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyObject_Call+0x43) [0x7fe17c063913]
<24>    Offset:         Address: 0x7fe17c0cde37 /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyEval_CallObjectWithKeywords+0x47) [0x7fe17c0cde37]
<25>    Offset:         Address: 0x7fe17c03c381 /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyEval_CallFunction+0xa1) [0x7fe17c03c381]
<26>    Offset:         Address: 0x7fe17ce1caf4 /usr/lib/irods/plugins/rule_engines/libirods_rule_engine_plugin-python.so(+0x765af4) [0x7fe17ce1caf4]
<27>    Offset:         Address: 0x7fe17ce0e1ca /usr/lib/irods/plugins/rule_engines/libirods_rule_engine_plugin-python.so(+0x7571ca) [0x7fe17ce0e1ca]
<28>    Offset: 0x834   Address: 0x7fe17ce0bb14 exec_rule(std::__1::tuple<> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::list<boost::any, std::__1::allocator<boost::any> >&, irods::callback)
<29>    Offset: 0x141   Address: 0x7fe17ced1b21 irods::error std::__1::__invoke_void_return_wrapper<irods::error>::__call<irods::error (*&)(std::__1::tuple<> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::list<boost::any, std::__1::allocator<boost::any> >&, irods::callback), std::__1::tuple<>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::list<boost::any, std::__1::allocator<boost::any> >&, irods::callback>(irods::error (*&&&)(std::__1::tuple<> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::list<boost::any, std::__1::allocator<boost::any> >&, irods::callback), std::__1::tuple<>&&&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&&&, std::__1::list<boost::any, std::__1::allocator<boost::any> >&&&, irods::callback&&)

<snip>
```

See the attached log for the full output: [prep_invalid_pointer_free.log.txt](https://github.com/irods/irods_rule_engine_plugin_python/files/7757826/prep_invalid_pointer_free.log.txt)

Further investigation will be needed, but it is definitely related to fixes introduced for leaks occurring which were discovered in irods/irods#4649.

In particular, the line causing the invalid pointer free - seen in frame <7> - is here, added in 4.2.11.0: https://github.com/irods/irods_rule_engine_plugin_python/blob/785b279cfc6e44d40b31c7bba78d2fe8593731a2/irods_rule_engine_plugin-python.cxx#L278

The issue seems to only manifest itself when a python exception is being thrown (notice frame <13>). When a proper data object path is given to the microservice, an error is not generated and the crash does not occur.

**Comment by trel** [comment]
The `clearMsParam()` fixes the 32k leak - so we want to keep it.  It allows for multiple rules to be run within the same connection without making the server leak until collapse.

So, wrapping line 278 in some logic to detect whether we're already 'in an error case' would prevent this pointer free.

More investigation to whether that means that a leak would still occur, but maybe that doesn't matter since it's already in an exception case.

Another consideration is when `!retVal.ok()` but when *not* caused by an exception (just a regular iRODS error code).

If an exception is thrown, but caught and handled in the rule itself... how would this behave?

@stsnel, I think we can get this done pretty quickly for a 4.2.11.1.

**Comment by stsnel** [comment]
@trel Thank you, that sounds like a great solution 👍

**Comment by alanking** [comment]
We have confirmed that the solution works and prevents leaks for non-error cases.

I did notice that some microservices still leak and we will try to address this in the next server release in addition to the other observed leaks. See irods/irods#6044 and irods/irods#6108
